### ISSUE TITLE
[Routing] Missing redirection when route param does not accept trailing slash

### ISSUE THREAD

**Issue by marc-romera** [symfony/symfony#30863]
**Symfony version(s) affected**: 4.2.4

**Description**  
Symfony 4 does not redirect in a route ending with a param which does not accept a trailing slash. Route can disallow a trailing slash if configured as shown here. Symfony 4 should detect that the route is invalid with a trailing slash (but valid without it) and so return a redirect response to the same route without the slash.

**How to reproduce**  
Step 1. Set your route

```
#app/config/routing.yml
index:
    path:       /check/{token}
    controller:  AppBundle:Default:token
    defaults:
        token: 1
    requirements:
        token: \d+
```

Step 2. Implement the action

```
// src/AppBundle/Controller/DefaultController.php
class DefaultController extends Controller
{
    /**
     * @param $token
     *
     * @return Response
     */
    public function tokenAction(string $token)
    {
        return new Response('token is ' . $token);
    }
}
```

Step 3. Serve the application and visit the path `http://127.0.0.1:8000/check/123/`
The output is the one it would provide without the slash:
`token is 123`

But we want it to actually be redirected to `http://127.0.0.1:8000/check/123`, which is the only valid one, already working and returning the same exact response.

**Additional context**  
Route-related log is identical *in both cases* (i.e., with and without the trailing slash):
`[2019-04-04 09:50:35] request.INFO: Matched route "test". {"route":"test","route_parameters":{"_route":"test","token":"123","_controller":"AppBundle\\Controller\\DefaultController::tokenAction"},"request_uri":"http://127.0.0.1:8000/check/123/","method":"GET"} []`

PS: Example heavily based on https://github.com/symfony/symfony/issues/29924, which seems somewhat related to this issue

**Comment by nicolas-grekas** [comment]
Related to #30013
The logic behind this is complex and I'm not sure your requirements don't conflict with some others expressed in the issues linked to this PR.

**Comment by marc-romera** [comment]
The related issue you provide allow trailing slashes, while the problem here is when these are not allowed, so requirements don't conflict.

One difference can be easily seen in the profiler. With my example the routing profiler says the path does not match (although it finally reinterprets it somehow and returns a 200):
![image](https://user-images.githubusercontent.com/18640998/55548840-b4ed0b00-56d4-11e9-861d-9fc0903dea74.png)

If I change the regex in the route definition to `.+` instead of `\d+`, then the route matches, and also the response changes to `token is 123/` instead of `token is 123`.

**In summary**: Your related issue and this one already behave different (routing treats them differently and even response is different), so they are different situations with different requirements, and this one behaves inconsistently

**Comment by nicolas-grekas** [comment]
Fixed in #31107
